Clarify specification (#12)
editorconfig/specification: index.rst

    modification, are permitted provided that the following conditions are met:

    1. Redistributions of source code must retain the above copyright notice,

>        this list of conditions and the following disclaimer.

Author: vscode was complaining about line length after i installed some rst detections

File Format
===========

> 

Author: @xuhdev I forgot to mention that I removed this reference to the Python configparser library so that this spec can stand on its own two feet.

Reviewer: Sounds good to me. Maybe we should delete them after we have included the AST?

Author: I’m not sure that it matters, as I’m not seeing any EC tests that actually support the many features of Python’s ConfigParser. There’s nothing breaking in removing this reference, as far as I can tell.

Author: Especially when you consider that all the non-Python cores are not using ConfigParser either.

Reviewer: Sounds good

EditorConfig uses an INI file format. In an EditorConfig file (usually named
``.editorconfig``), all beginning whitespace on each line is considered

> irrelevant. Each line must be one of the following:

Author: Irrelevant is better than ignored, as the parser could very well included this whitespace in the AST for reproducing the document as it was originally.

Reviewer 2: :thought_balloon: not only "an INI" file format but the strict editorconfig-INI file format?

Author: I will clarify. Thanks!

> - Blank: contains only whitespace characters.

Author: Blank is better than empty, because empty is traditionally used to indicate a sring with no contents, including whitespace. Blank may include whitespace characters.

Reviewer: Right. good catch!

- Comment: starts with a ``;`` or a ``#``.

>    - Inline comments are not supported.

Author: I moved this point as a sub-point of the `Comment` bullet here so that the reader has more context. I continued this paradigm throughout the PR.

Reviewer: I feel "supported" is a somewhat ambiguous word.  Plus the definition of "Inline comments" is gone.

Author: How about this?

> Inserting a `#` or `;` after non-whitespace characters in a line (i.e., inline) will not be parsed as a comment; in fact, it will be considered part of the section name or property in which you inserted it, which might not be what you intended.

Reviewer: I think we have left this undefined as in https://github.com/editorconfig/editorconfig-vote/issues/6 . How about this?

> Inserting a `#` or `;` after non-whitespace characters in a line (i.e., inline) is undefined at this moment but reserved.

Author: What does undefined mean, exactly? How would that parse into AST, for example?

I'm left puzzled by [the remaining two tests](https://github.com/editorconfig/editorconfig-core-test/blob/48610d43b7455af12195473377f93c4ceea654f5/parser/comments.in#L47-L53):

```ini
# Octothorpe at end of value read as part of value
[test11.c]
key=value# not comment
```
☝️ notice how GitHub grays out the inline comment.

If `# not comment` is not a comment, then what is it? Undefined? Can we please define undefined in this context? Also, since we're trimming keys and values, would the value of `foo = bar # baz` be `bar` or `bar ` with a space before the comment indicator? We really need to 🔒 this down!

```ini
# Escaped octothorpe in value
[test12.c]
key=value \# not comment
```
☝️ notice again how GitHub grays this one out too, confusingly.

Reviewer 2: I remember i've read parser-descriptions calling lines starting with a "comment"-character being ignored.

In that case the AST just wouldn't contain them as well as blank lines. Having ignored lines in the AST allows further use cases, like comment-linting or even more sophisiticated use cases (semantic use of comments like f.e. macros). This is probably not inteded for ecfg but for an INI-file in general it may be interesting?

Author: Unfortunately, there is no way to reproduce the original document without saving these values in the AST. Quite technically, what you are asking for is a concrete syntax tree (CST), which does not contain “irrelevant” content, including blank lines and comments.

Do you propose that we have both an AST as well as a CST representation of the document?

Author: Also, I’m not sure that being “ignored” means it doesn’t become part of the AST. It sounds to me like it means the implementation of that AST are just ignoring those values, which is consistent with this specification as well.

Reviewer 2: Oh and sorry for the confusing reply: We shouldn't really do not recognize inline comments: Different Ini-Parsers may behave differently since in "real" Ini-files there might be values containing octothorpes?

Reviewer 2: > Also, I’m not sure that being “ignored” means it doesn’t become part of the AST. It sounds to me like it means the implementation of that AST are just ignoring those values, which is consistent with this specification as well.

Good point - i thought we would anyways keep everything "we don't use" in the AST. So if somebody else want's to implement inline-comments he can do easily on top?

Author: Correct. Different INI passers definitely do different things. I don’t think we support inline comments, so we have to decide what to do with inline comment indicators and I’m still getting mixed or confusing feedback on this one item.

Author: The confusing part is that we’re saying we don’t support inline comments; yet, we’re saying in the same breath that if you do it it will be ignored? That sounds an awful lot like we’re supporting them, because ignoring what comes after a comment indicator is exactly the definition of a comment!

Reviewer 2: Ahh - i see. I just implicitly thought, if the line is not introduced by a significant character (`# ; [`) it will be handled like a normal key-value line. Saying the inline-comments will become part of the value.

Since this might be a breaking change - we should

1. detect this in the transition-phase of editiorconfig-spec (and handle it accordingly after the AST-generation) and throw a warning **or**
2. allow that in the editorconfig-spec in general (and handle it accordingly after the AST-generation) **or**
3. allow it for editorconfig-ini and handle it in the AST-generation.

---
I hope you can handle that my statements iterate with my level of understanding..

Reviewer 2: Oh and i 'd favor option 1 or 2

Author: The current EC tests, if I understand correctly, do not allow inline comments; thus, treat them as part of the key or value, depending on where you put them. That's different than being ignored though, isn't it?

Again, I don't think we need to introduce a breaking change to add clarity.

As such, I personally think we should be backwards compatible and clarify that:

> Inserting a `#` or `;` after non-whitespace characters in a line (i.e., inline) will not be parsed as a comment; in fact, it will be considered part of the section name or key-value pair in which it was inserted, which might not be the author's intent.

Comments should be handled in the formal grammar, which ends up in the AST.

If we do want to support inline comments, that would be a breaking change and we'd need to create an issue about that. For now, I just want to add clarity to the existing spec as it relates to the existing EC tests.

Reviewer 2: It's difficult for me to argument the right way - i have the feeling i'm mixing up the EC-INI-spec with the EC-spec. Would it be heretic thinking about creating an EC-INI-parser creating an AST and using the EC-parser to transform into a EC-specific variant?

Reviewer 2: Like having `[INI-file] -> [INI-AST] -> [EC-AST]`?

Reviewer 2: Besides of that do i have the strong need to make a tabular comparison of different INI-parsers (which is probably not needed for the EC-spec on its own). And if we decide to go for an explicit EC-INI-parser, too, i'll try to collect some INI-stats as well.

Author: I don't think that makes sense, because EditorConfig is intended to be flexible in that it supports any custom declarations you throw at it. Let me try to help you draw the line between the AST and EC's logic though:

## INI Parsing into AST

The INI parser just handles the parsing of the INI file, strictly according to the INI specification. In theory, it wouldn't need to know anything about EditorConfig, because it doesn't have any business logic in it to that effect. This means the INI AST can be used by anyone in contexts that have nothing to do with EditorConfig.

## EditorConfig Business Logic

- Merging INI files until `root=true` is found (the final merge might be the EC AST you were talking about).
- Parsing the section headers as filepath globs.
- Ignoring declarations that are not directly supported by EditorConfig, even though they exist in the AST.
- In the future, throwing validation errors that are specific to EditorConfig and not to the INI parsing (e.g., `indent_size` cannot equal `two` and expected an integer value).

Reviewer 2: I see - thanks for the clarification! It's hard for me to create an opinion about something as controversial¹ as inline-comments.

¹) I snooped into some INI-parser-implementations and many of them treat the different or provide comment-specific modes. It seems to be the same with value-parsing, though. There seem to exist different parsers allowing to encapsulate the value in double-quotes allowing to have significants in the value - others seem to have not.

For me this is at first a INI-grammar-specific problem.

Reviewer: According to editorconfig/editorconfig-vote#6 , I think we have left the decision of this task to a future date. We neither treat text after a semicolon or sharp as part of the value nor comment. The rationale is that we haven't encountered a situation in which we should make a decision with respect to this issue. It likely leaves us more flexibility and leeway when we need to make a decision because there is a need. This is also pretty common strategy in the development of specs in other areas. For now, we can just say, users should not use the so-called "inline comments". We do not guarantee what will happen at this point.

Reviewer: > I'm left puzzled by [the remaining two tests](https://github.com/editorconfig/editorconfig-core-test/blob/48610d43b7455af12195473377f93c4ceea654f5/parser/comments.in#L47-L53):
> 
> ```ini
> # Octothorpe at end of value read as part of value
> [test11.c]
> key=value# not comment
> ```
> 
> point_up notice how GitHub grays out the inline comment.
> 
> If `# not comment` is not a comment, then what is it? Undefined? Can we please define undefined in this context? Also, since we're trimming keys and values, would the value of `foo = bar # baz` be `bar` or `bar ` with a space before the comment indicator? We really need to lock this down!
> 
> ```ini
> # Escaped octothorpe in value
> [test12.c]
> key=value \# not comment
> ```
> 
> point_up notice again how GitHub grays this one out too, confusingly.

These two tests are probably something we forgot to remove after https://github.com/editorconfig/editorconfig-vote/issues/6 . They are testing what https://github.com/editorconfig/editorconfig-vote/issues/6 exactly opposes to.

I don't think GitHub grays it out means anything (plus its buggy syntax highlighting). People define INI differently. For example, [configparser](https://docs.python.org/3/library/configparser.html) says:

> Configuration files may include comments, prefixed by specific characters (# and ; by default 1). **Comments may appear on their own on an otherwise empty line, possibly indented.**

"Undefined" might be better named "implementation-defined": "No warranty, users should avoid using it. How a so-called 'inline comment' is handled is not uniform across implementation at this moment."  As commented above, this is to leave more flexibility when we truly need to make a decision about this issue for another new feature in the future.

Author: Resolved in https://github.com/editorconfig/specification/pull/12/commits/3d96e629245a53844b1cd64611cfdf054265c9e7.

- Section Header: starts with a ``[`` and ends with a ``]``.
   - May not use any non-whitespace characters outside of the surrounding
     brackets.
   - May contain any characters between the square brackets (e.g.,
     ``[`` and ``]`` are allowed).
   - Forward slashes (``/``) are used as path separators.
   - Backslashes (``\\``) are not allowed as path separators (even on Windows).
   - Section Name: the string between the beginning ``[`` and the ending ``]``.
   - Section: the lines starting from a Section Header until the beginning of the

>      next Section Header or the end of the file.

Reviewer: Can we say these lines are definitions? It might confuse readers.

Author: But they aren't definitions. They could be comments, blank lines or properties (key-value pairs).

Reviewer: What do you mean?

> Section Name: the string between the beginning ``[`` and the ending ``]``.

looks like a classic definition to me.

Author: What? You made this comment on the "Section" line, not the "Section Name" line. Please clarify.

Reviewer: The comment is over the 3 lines (i.e., the last two bullet point). I'm not sure how to tell that from GitHub web interface...

Author: Hmm. They are nested here because they are closer to the relevant context, but I agree it’s a bit strange. I can’t unnest them, however, as they do not relate to individual lines in an INI file.

I’ll find a good place to move them.

> - Property: contains a key-value pair, separated by an `=`.

Author: Renamed this from Assignment to Property for consistency with the rest of the document as well as the Wikipedia article about the INI file format.

Reviewer: From Wikipedia, Does "Property" mean the left-hand side of the key-value pair, i.e., the key?

Author: No, the Wikipedia article always refers to the property as the whole thing. Unfortunately, it uses key and property interchangeably for the same thing, which is not ideal.

The YML specification refers to them as [key: value pairs](https://yaml.org/spec/1.2/spec.html#key:%20value%20pair//) and so does JSON, JavaScript and many other languages.

Hmm... it looks like CSS refers to each [property declaration as a name and value](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleDeclaration/setProperty).

I'm actually OK renaming "key" back to "name" if that's more clear. It's just not a consistent nomenclature throughout languages, unfortunately.

Reviewer: I see [CSS spec](https://www.w3.org/TR/2019/CR-css-syntax-3-20190716/) uses "property" loosely to refer to either the left-hand side or the whole assignment. "Property" seems to me is sort of loose. Given that "property" is used loosely else where, I'm not sure it's worth to abandon the current terminology here; we probably have a lot of wiki pages and there are other online resources that have already followed the current definition.

Reviewer: [Yaml spec](https://yaml.org/spec/1.2/spec.html) also seems like so:

>  Both the node’s properties and node content are optional.

Author: That's not being referred to as the left side. Where do you see it being referred to as the left side? I don't think Property is loose at all. It _always_ refers to the whole thing, where I've seen it used. When the CSS spec says "color property" don't confuse that to mean that "color" is itself a property. The (color) property is a property with a key named color and a hex value. It's still a property, but they are using the key of the property as a sort of label so you know the property to which they are referring.

Author: My name is Jed Mao. If you saw 100 names in a file, you might say, "which line is Jed's name on?" Same thing. Which line is the color's property on?

Reviewer: Here's the relevant quote from a highlighted piece of text from the above CSS spec doc:

> CSS renderers must treat as invalid (and ignore as appropriate) any at-rules, properties, property values, keywords, and other syntactic constructs for which they have no usable level of support.

I don't see a clear meaning of "property" in its other occurrences in that doc, but in this sentence, I think "properties" more likely means the left-hand side than the whole thing.

Reviewer: I looked up [JSON spec](https://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf) as well, and there is no mention of "property." Right now I've had no luck in searching yaml, CSS, and JSON specs for a clear definition of "property" as used in a similar way as we desire to use. If the word "property" is so vague and others are avoiding using it, I think it might be easier to just avoid using this word, and let existing online resources refer to whatever it used to mean, maybe loosely, maybe more precisely. Alternatively we can use word like "declaration" for key-value pair, "key" for the left-hand side, "value" for the right-hand side. What do you think?

Author: I’m OK with that. I got “declaration” [from PostCSS](https://github.com/postcss/postcss/blob/master/lib/declaration.es6).

Reviewer: OK. Shall we go for "declaration"? (BTW "statement" might also be a good candidate)

Author: Yeah. I think we should stick with declaration. It also seems to be consistent with [the description of the const declaration in JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const#description).

>    - Key: the part before the first `=` (trimmed of whitespace).

Author: Renamed this from Name to Key, as that is more consistent with modern nomenclature.

File Format
===========

> EditorConfig uses a strict EditorConfig-INI file format. In an EditorConfig

Author: That was @florianb's suggestion, because now the INI file format specification is locked down with the additional clarification we need. It no longer leans on the 3rd party Python ConfigParser specification. It stands on its own, strictly.

Why do you think we don't need this? What is your definition of strict that wouldn't apply here?

> EditorConfig uses a strict EditorConfig-INI file format. In an EditorConfig

Reviewer 2: In future iterations this also should reference to `editorconfig/ini` to make it even more clear.

> EditorConfig uses a strict EditorConfig-INI file format. In an EditorConfig

Reviewer: This is only a language suggestion: I think this is a bit redundant and pretty much every concept is strict in this document. It is also "strictly" UTF-8 encoded, keys are strictly to the left of `=`, etc.

> EditorConfig uses a strict EditorConfig-INI file format. In an EditorConfig

Author: Removed strict in 605c0c4

file (usually named ``.editorconfig``), all beginning whitespace on each line
is considered irrelevant. Each line must be one of the following:

- Blank: contains only whitespace characters.
- Comment: starts with a ``;`` or a ``#``.
   - Inserting a ``#`` or ``;`` after non-whitespace characters in a line
     (i.e., inline) shall neither be parsed as a comment nor as part of the
     section name, key or value in which it was inserted. This may change in

>      the future.

Reviewer: ```suggestion
     the future. Users should avoid this uncertain syntax.
```

Reviewer 2: I wouldn't spend too much time on that point, yet. As soon as we get a spec for the ini-file having potentially a roadmap affecting this eventually it will be by far easier to deal with changes to current behavior. My personal target for the first iteration would be to have a fundamental spec to build on.

Author: We've already spent a lot of time getting here!

Reviewer 2: True but adding "users should avoid this uncertain syntax" adds implicitly that we already know where we go with that point, yet. And as far as i understood it we do not. And i'd like to avoid starting that discussion yet.

Reviewer 2: But i also have no issues with it - i was fine withe last version as well.

Reviewer: > True but adding "users should avoid this uncertain syntax" adds implicitly that we already know where we go with that point, yet. And as far as i understood it we do not. And i'd like to avoid starting that discussion yet.

I thought this would clarify that we don't know what will happen :) Because we do not know what will happen, we kindly ask our users to not engage in doing so.

- Section Header: starts with a ``[`` and ends with a ``]``.
   - May not use any non-whitespace characters outside of the surrounding
     brackets.
   - May contain any characters between the square brackets (e.g.,
     ``[`` and ``]`` are allowed).
   - Forward slashes (``/``) are used as path separators.
   - Backslashes (``\\``) are not allowed as path separators (even on Windows).
- Declaration: contains a key-value pair, separated by an `=`.
   - Key: the part before the first `=` (trimmed of whitespace).
   - Value: The part after the first `=` (trimmed of whitespace).

Any line that is not one of the above is invalid.

EditorConfig files should be UTF-8 encoded, with LF or CRLF line separators.

Additionaly, EditorConfig defines the following terms:

> - Preamble: the part of the document that precedes the first section. The

Reviewer: ```suggestion
- Preamble: the lines that precede the first section. The
```

> Supported Properties

Reviewer: ```suggestion
Supported Keys
```

Author: I think this should be "Supported Declarations" because this section of the document does cover supported keys, but also which values are supported for those keys; thus, the entire declaration support.

Reviewer: Sounds good

====================

> EditorConfig file sections contain properties, which are key-value pairs

Reviewer: "properties, which are key-value pairs ...." --> "declarations"?

Reviewer 2: I'd prefer easier language here. Not everybody managing editorconfig-files may be a developer?

Author: Do you have a suggestion?

Reviewer 2: Key-Value is fine for me - the Wikipedia-css page talks about "property"/"value"-pairs. I just would probably avoid "declaration" even though its technically the precise term it is probably needlessly complicated.

Reviewer: I think it is important to use more precise language in a spec. How about employing some tautology?

> ... declarations, i.e., key-value pairs ...

     - Set to ``latin1``, ``utf-8``, ``utf-8-bom``, ``utf-16be`` or ``utf-16le`` to
       control the character set. Use of ``utf-8-bom`` is discouraged.
   * - ``trim_trailing_whitespace``
     - Set to ``true`` to remove all whitespace characters preceeding newline
       characters in the file and ``false`` to ensure it doesn't.
   * - ``insert_final_newline``
     - Set to ``true`` ensure file ends with a newline when saving and ``false``
       to ensure it doesn't.
   * - ``root``

>      - Must be specified in the default (global/root) section of the file. Set

Reviewer: We have defined section as "the lines starting from a Section Header until the beginning of the next Section Header or the end of the file." That's why I think it should be "outside of any sections." The statement before the change is also consistent with "Any property other than ``root`` MUST be
located under a section to take effect."

Author: Good point. The root section is still considered a section though. I've seen it referred to as the default section.

What do you think if we go the other way around and instead redefine a section to be:

> A collection of properties, comments and blank lines; typically, including a section header; the only exception being the default (i.e., global/root) section, which _never_ includes a section header.

I'd be happy to create a graphic to illustrate "The Anatomy of an EditorConfig INI File" in order to suppliment this documentation and add even more clarity.

Reviewer: Since we have mentioned so many times of section within the document, do you think it's just easier to call it something different, such as preamble?

> Preamble: From the beginning to the first Section Header.

Author: I was going to say no, but the more I look into the AST interfaces, that does actually make a lot of sense.

## INI parsing

The preamble and the collection of sections that follow:

- Are siblings!
- May contain blank lines, comments and any number of properties.

## EditorConfig-Specific Rules

- The preamble may only contain a single property of `root=true`.

Reviewer 2: Uhm - i am sure i missed that but i use the preamble to define "global" defaults. Is that considered being bad behaviour and should be done below an explicit asterisk-section?

Author: Can you give me an example INI file for context?

Reviewer 2: No i can't - after crawling through my ecfg-files i found out i consistently use the `[*]`-section. I shouldn't write comments while being distracted.. Sorry. :boy: :girl:

Author: All good, my friend.

     - Set to ``latin1``, ``utf-8``, ``utf-8-bom``, ``utf-16be`` or ``utf-16le`` to
       control the character set. Use of ``utf-8-bom`` is discouraged.
   * - ``trim_trailing_whitespace``
     - Set to ``true`` to remove all whitespace characters preceeding newline
       characters in the file and ``false`` to ensure it doesn't.
   * - ``insert_final_newline``
     - Set to ``true`` ensure file ends with a newline when saving and ``false``
       to ensure it doesn't.
   * - ``root``
     - Must be specified in the preamble. Set to ``true`` to stop the
       ``.editorconfig`` file search on the current file. The value is case
       insensitive.

> For any property, a value of ``unset`` removes the effect of that

Reviewer: ```suggestion
For any declaration, a value of ``unset`` removes the effect of that
```

property, even if it has been set before. For example, add ``indent_size =
unset`` to undefine the ``indent_size`` property (and use editor defaults).

Property names are case insensitive and all property names are lowercased when
parsing. The maximum length of a property name is 50 characters and the

> maximum length of a property value is 255 characters. Any property beyond

Reviewer: ```suggestion
maximum length of a property value is 255 characters. Any key and value beyond
```
